[llvm/Support/PrefixMapper] Refactorings for `PrefixMapper`

akyrtzi opened the pull request:
* Unify `PrefixMapper` and `TreePathPrefixMapper`; make `PrefixMapper` the canonical interface and `TreePathPrefixMapper` a subclass implementation of it that overrides the `add` and `mapImpl` functions.
* Remove `StringSaver` from `PrefixMapper`. I find the fact that it would implicitly allocate memory every time you called `map` undesirable, and returning a `StringRef` was not a commonly needed property anyway.
* Streamline it a bit and remove some APIs that were not used beyond unit tests.

This is a step towards using `PrefixMapper` for include-tree, which doesn't use a `CachingOnDiskFileSystem`.

akyrtzi committed:
[llvm/Support/PrefixMapper] Refactorings for `PrefixMapper`

* Unify `PrefixMapper` and `TreePathPrefixMapper`; make `PrefixMapper` the canonical interface and
  `TreePathPrefixMapper` a subclass implementation of it that overrides the `add` and `mapImpl` functions.
* Remove `StringSaver` from `PrefixMapper`. I find the fact that it would implicitly allocate memory every time you called `map`
  undesirable, and returning a `StringRef` was not a commonly needed property anyway.
* Streamline it a bit and remove some APIs that were not used beyond unit tests.

This is a step towards using `PrefixMapper` for include-tree, which doesn't use a `CachingOnDiskFileSystem`.

akyrtzi commented:
@swift-ci Please test

benlangmuir commented on llvm/unittests/Support/PrefixMapperTest.cpp:
```diff
@@ -114,8 +136,8 @@ TEST(PrefixMapperTest, construct) {
   PrefixMapper PM(sys::path::Style::posix);
-  PM.add(MappedPrefix{"a", "b"});
-  PM.add(MappedPrefix{"b", "a"});
+  cantFail(PM.add(MappedPrefix{"a", "b"}));
```
Should we use `ASSERT_THAT_ERROR(..., Succeeded())` for tests?  `cantFail` is only an `assert` so it could fail silently in a no-asserts build.

akyrtzi replied:
Done 👍

benlangmuir commented on llvm/lib/TableGen/Main.cpp:
```diff
@@ -118,9 +118,11 @@ template <class StringSetT>
+    for (StringRef DepRef : Dependencies) {
+      std::string Dep = DepRef.str();
+      if (CreateDependencyFilePM)
+        cantFail(CreateDependencyFilePM->mapInPlace(Dep));
```
Why is this `cantFail`? Same for the one below.  We should either report the error or have a comment why it cannot fail when it's not obvious from immediate context.

akyrtzi replied:
It's because it creates a `PrefixMapper` instance that doesn't error, but I changed it to propagate a possible error, to be safe for future changes.

benlangmuir commented on llvm/lib/Support/PrefixMapper.cpp:
```diff
@@ -278,6 +222,15 @@ Error TreePathPrefixMapper::add(const MappedPrefix &Mapping) {
+  StringRef TreePath = Entry.getTreePath();
+  SmallString<256> PathBuf;
+  Optional<StringRef> Mapped =
+      cantFail(PrefixMapper::mapImpl(TreePath, PathBuf));
```
Why is this one `cantFail` but the call to `PrefixMapper::mapImpl`  in `TreePathPrefixMapper::mapImpl` is not?

akyrtzi replied:
I made both `cantFail` for consistency.

benlangmuir reviewed (commented):
A few `cantFail` questions, but otherwise LGTM.

akyrtzi commented:
@swift-ci Please test

akyrtzi committed:
[llvm/Support/PrefixMapper] Make prefix mapper own the storage for the mappings

The mappings are commonly going to only be a handful, simplify use of the prefix mapper by changing the mappings
to use `std::string` so as to avoid worrying about managing lifetimes of `StringRef`s.

akyrtzi commented:
@swift-ci Please test

benlangmuir commented on llvm/include/llvm/Support/PrefixMapper.h:
```diff
@@ -24,8 +24,12 @@ class CachedDirectoryEntry;
+  std::string New;
+
+  MappedPrefix() = default;
+  MappedPrefix(StringRef Old_, StringRef New_)
```
Nit: you don't need the _ in a constructor, since `X(X)` already does what you want.

akyrtzi replied:
I don't like naming parameters in constructor same as the member, it is a footgun that has bitten me before.

benlangmuir replied:
I appreciate the potential problem, but you're fighting against the dominant style in llvm.
* `X(X)` 6864 hits (regex `[:,] ([A-Za-z0-9_]+)\(\1\)'`)
* `X(X_)` 295 hits (regex `[:,] ([A-Za-z0-9_]+)\(\1_\)'`)
* `X(_X)` 105 hits (regex `[:,] ([A-Za-z0-9_]+)\(\1_\)'`)

I didn't exhaustively confirm those hits come from constructor params, and all of these will undercount a bit, but I think the trend is fairly obvious.

Anyway, feel free to land it either way, it doesn't bother me it just stands out.

akyrtzi replied:
AFAICT https://llvm.org/docs/CodingStandards.html doesn't forbid this scheme 🤷🏻‍♂️
